engine/src/main/java/org/pentaho/di/trans/steps/regexeval/RegexEval.java

```diff
@@ -0,0 +1,268 @@
+        logError( BaseMessages.getString( PKG, "RegexEval.Log.ErrorFindingField" ) + "[" + meta.getMatcher() + "]" );
+        throw new KettleStepException( BaseMessages.getString( PKG, "RegexEval.Exception.CouldnotFindField", meta
+          .getMatcher() ) );
+      }
+
+      // Cache the position of the CaptureGroups
+      if ( meta.isAllowCaptureGroupsFlagSet() ) {
+        data.positions = new int[meta.getFieldName().length];
+        String[] fieldName = meta.getFieldName();
+        for ( int i = 0; i < fieldName.length; i++ ) {
+          if ( fieldName[i] == null || fieldName[i].length() == 0 ) {
+            continue;
+          }
+          if ( meta.isReplacefields() ) {
+            data.positions[i] = data.outputRowMeta.indexOfValue( fieldName[i] );
+          } else {
+            data.positions[i] = captureIndex;
+            captureIndex++;
+          }
+        }
+      } else {
+        data.positions = new int[0];
+      }
+
+      // Now create objects to do string to data type conversion...
+      //
+      data.conversionRowMeta = data.outputRowMeta.cloneToType( ValueMetaInterface.TYPE_STRING );
+    }
+
+    // reserve room
+    Object[] outputRow = RowDataUtil.allocateRowData( data.outputRowMeta.size() );
+    System.arraycopy( row, 0, outputRow, 0, getInputRowMeta().size() );
+
+    try {
+      // Get the Field value
+      String fieldValue;
+      boolean isMatch;
+
+      if ( getInputRowMeta().isNull( row, data.indexOfFieldToEvaluate ) ) {
+        fieldValue = "";
+        isMatch = false;
+      } else {
+        fieldValue = getInputRowMeta().getString( row, data.indexOfFieldToEvaluate );
+
+        // Start search engine
+        Matcher m = data.pattern.matcher( fieldValue );
+        isMatch = m.matches();
+
+        if ( meta.isAllowCaptureGroupsFlagSet() && data.positions.length != m.groupCount() ) {
+          // Runtime exception case. The number of capture groups in the
+          // regex doesn't match the number of fields.
+          logError( BaseMessages.getString( PKG, "RegexEval.Log.ErrorCaptureGroupFieldsMismatch", String
+            .valueOf( m.groupCount() ), String.valueOf( data.positions.length ) ) );
+          throw new KettleStepException( BaseMessages.getString(
+            PKG, "RegexEval.Exception.ErrorCaptureGroupFieldsMismatch", String.valueOf( m.groupCount() ), String
+              .valueOf( data.positions.length ) ) );
+        }
+
+        for ( int i = 0; i < data.positions.length; i++ ) {
+          int index = data.positions[i];
+          String value;
+          if ( isMatch ) {
+            value = m.group( i + 1 );
+          } else {
+            value = null;
+          }
+
+          // this part (or possibly the whole) of the regex didn't match
+          // preserve the incoming data, but allow for "trim type", etc.
+          if ( value == null ) {
+            try {
+              value = data.outputRowMeta.getString( outputRow, index );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

On this line, index and outputRow are used as input of method call getString() to compute value. If this method call throws an exception, the value of value is not updated, thus, the states of index and outputRow, and value might be inconsistent when used later. Consider verifying the value of value after calling getString() to ensure the value is what you expect it to be.